File: languages/python/src/operations/quick_sort.py
"""
Contains implementation of QuickSort.
"""

def quick_sort(arr: list[int], low: int, high: int):
    if low < high:
        pivot_index = partition(arr, low, high)
        quick_sort(arr, low, pivot_index - 1)
        quick_sort(arr, pivot_index + 1, high)

    return arr

def partition(arr: list[int], low: int, high: int) -> int:
    pivot = high     # Last element is pivot
    i = low - 1     # Index of smaller element

    for j in range(low, high):
        if arr[j] < arr[pivot]:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    i += 1      
    arr[i], arr[pivot] = arr[pivot], arr[i]
    return i

File: languages/c#/src/operations/QuickSort.cs
namespace Operations
{
    public static class QuickSortOperation
    {
        private static int Partition(IComparable[] arr, int low, int high)
        {
            int pivot = high;   // Last element is pivot
            int i = low - 1;    // Index of smaller element

            for (int j = low; j < high; j++)
            {
                if (arr[j].CompareTo(arr[pivot]) < 0)
                {
                    i++;
                    (arr[i], arr[j]) = (arr[j], arr[i]);
                }
            }

            i++;
            (arr[i], arr[pivot]) = (arr[pivot], arr[i]);
            return i;
        }

        private static IComparable[] QuickSort(IComparable[] arr, int low, int high)
        {
            if (low < high)
            {
                int pivotIndex = Partition(arr, low, high);
                QuickSort(arr, low, pivotIndex - 1);
                QuickSort(arr, pivotIndex + 1, high);
            }
            return arr;
        }
        
        public static IComparable[] QuickSort(IComparable[] arr)
        {
            return QuickSort(arr, 0, arr.Length - 1);
        }
    }
}
